Whole-string match reqs in comment_requirement

TheBassEngineer opened the pull request:
Change `comment_requirement(req)` to look for whole-string matches to module names listed in COMMENT_REQUIREMENTS.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Proposal to fix #55178 by making `comment_requirements` look for exact, whole-string matches. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #55178
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

TheBassEngineer committed:
Whole-string match reqs in comment_requirement

Change `comment_requirement(req)` to look for whole-string matches to module names listed in COMMENT_REQUIREMENTS.

frenck commented on script/gen_requirements_all.py:
```diff
@@ -154,7 +154,10 @@ def gather_recursive_requirements(domain, seen=None):
 def comment_requirement(req):
     """Comment out requirement. Some don't install on all systems."""
-    return any(ign.lower() in req.lower() for ign in COMMENT_REQUIREMENTS)
+    idx=req.find("==")
+    if idx >= 0:
+      return any(ign.lower() == req[0:idx].lower() for ign in COMMENT_REQUIREMENTS)
+    return any(ign.lower() == req.lower() for ign in COMMENT_REQUIREMENTS)
```
Why not use:

```suggestion
    return any(f"{ign.lower()}==" in req.lower() for ign in COMMENT_REQUIREMENTS)
```

TheBassEngineer replied:
I'm not actually sure if package manifests are allowed to use `module>=version` (although my code still doesn't handle that case correctly either).  If we can depend on it always being `module==version`, then yes that is 100% better.

ludeeus replied:
We only allow `==`

TheBassEngineer replied:
Perfect. Accepting the suggestion now.

TheBassEngineer committed:
Update script/gen_requirements_all.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

TheBassEngineer committed:
Use regex to correctly match reqs with extras

If I've done my regex homework right, the proposed regex in `comment_requirement` should also now correctly match requirement strings with an optional bracketed extra, of the form `module[extra]==x.x`, so that `beacontools` in COMMENT_REQUIREMENTS still correctly matches `beacontools[scan]=1.2.3` but `decora` doesn't match `decora_wifi==1.4`.

I can't deny it's kinda ugly looking though.

I can't deny it's ugly though.

TheBassEngineer committed:
Revert "Use regex to correctly match reqs with extras"

This reverts commit 437c23731c74a31596bb49e73bb60e6dd78dfe44.

TheBassEngineer committed:
Add beacontools[scan] to COMMENT_REQUIREMENTS

TheBassEngineer committed:
Remove redundant beacontools from COMMENT_REQUIREMENTS

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

TheBassEngineer committed:
Use `startswith` in the check.

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

TheBassEngineer committed:
Make check case-sensitive.

TheBassEngineer committed:
Use hassfest pkg name normalization logic

TheBassEngineer commented:
Had to move `normalize_package_name` from `hassfest/requirements` into `gen_requirements_all` to avoid a circular dependency. Let me know if there is a helper library that both could reference which would be a better home for it.
